File: apps/backend-node/src/features/bookings/booking.repository.ts
import { Injectable } from '@nestjs/common';
import {
  booking_mode,
  booking_status,
  booking_type,
  type bookings,
  Prisma,
  property_status,
  room_status,
} from '@prisma/client';

import { PrismaService } from '../../prisma/prisma.service';
import {
  BadRequestError,
  ConflictError,
  NotFoundError,
} from '../../shared/errors';
import type { BookingPrice } from './booking-pricing.service';
import { rethrowBookingOverlap } from './booking-overlap.util';

type TransactionClient = Prisma.TransactionClient;

export interface BookingInputs {
  roomId: string;
  propertyId: string;
  bookingMode: string;
  minStayNights: number;
  monthlyPriceKrw: number;
  maxOccupancy: number;
  nights: number;
}

export interface BookingListRow {
  id: string;
  customerId: string;
  status: string;
  bookingType: string;
  checkIn: Date;
  checkOut: Date;
  guestCount: number;
  holdExpiresAt: Date | null;
  createdAt: Date;
  customerNotes: string | null;
  propertyId: string;
  roomId: string;
  propertyTitle: string;
  district: string;
  roomName: string;
  rentKrw: number;
  serviceFeeKrw: number;
  totalKrw: number;
  pricingVersion: string;
}

interface BookingListRowRecord {
  id: string;
  customer_id: string;
  status: string;
  booking_type: string;
  check_in: Date;
  check_out: Date;
  guest_count: number;
  hold_expires_at: Date | null;
  created_at: Date;
  customer_notes: string | null;
  property_id: string;
  room_id: string;
  property_title: string;
  district: string;
  room_name: string;
  rent_krw: number;
  service_fee_krw: number;
  total_krw: number;
  pricing_version: string;
}

const BOOKING_LIST_SQL = `
  select
    b.id,
    b.customer_id,
    b.status::text as status,
    b.booking_type::text as booking_type,
    b.check_in,
    b.check_out,
    b.guest_count,
    b.hold_expires_at,
    b.created_at,
    b.customer_notes,
    b.property_id,
    b.room_id,
    p.title as property_title,
    p.district,
    r.name as room_name,
    s.rent_krw,
    s.service_fee_krw,
    s.total_krw,
    s.pricing_version
  from public.bookings b
  join public.properties p on p.id = b.property_id
  join public.rooms r on r.id = b.room_id
  join public.booking_price_snapshots s on s.booking_id = b.id
`;

@Injectable()
export class BookingRepository {
  constructor(private readonly prisma: PrismaService) {}

  async validateBookingInputs(
    roomId: string,
    checkIn: Date,
    checkOut: Date,
    guestCount: number,
    tx: TransactionClient = this.prisma,
  ): Promise<BookingInputs> {
    if (checkOut <= checkIn) {
      throw new BadRequestError('checkOut must be after checkIn');
    }

    if (guestCount <= 0) {
      throw new BadRequestError('guestCount must be positive');
    }

    const room = await tx.rooms.findFirst({
      where: {
        id: roomId,
        deleted_at: null,
        status: room_status.available,
      },
    });

    if (!room) {
      throw new NotFoundError('Room is not available');
    }

    const propertyRow = await tx.properties.findFirst({
      where: {
        id: room.property_id,
        deleted_at: null,
        status: property_status.published,
      },
    });

    if (!propertyRow) {
      throw new NotFoundError('Property is not published');
    }

    const nights = daysBetween(checkIn, checkOut);

    if (nights < propertyRow.min_stay_nights) {
      throw new BadRequestError(
        `Stay must be at least ${propertyRow.min_stay_nights} nights`,
      );
    }

    if (guestCount > room.max_occupancy) {
      throw new BadRequestError(
        `Room supports up to ${room.max_occupancy} guests`,
      );
    }

    if (
      room.available_from !== null &&
      room.available_from > checkIn
    ) {
      throw new BadRequestError(
        'Room is not available from the selected check-in date',
      );
    }

    return {
      roomId: room.id,
      propertyId: propertyRow.id,
      bookingMode: propertyRow.booking_mode,
      minStayNights: propertyRow.min_stay_nights,
      monthlyPriceKrw: room.monthly_price_krw,
      maxOccupancy: room.max_occupancy,
      nights,
    };
  }

  async roomHasBookingConflict(
    roomId: string,
    checkIn: Date,
    checkOut: Date,
    tx: TransactionClient = this.prisma,
  ): Promise<boolean> {
    const rows = await tx.$queryRaw<{ exists: boolean }[]>`
      select exists (
        select 1
        from public.bookings b
        where b.room_id = ${roomId}::uuid
          and b.status in ('pending_payment', 'confirmed', 'active')
          and daterange(b.check_in, b.check_out, '[)')
              && daterange(${checkIn}::date, ${checkOut}::date, '[)')
      ) as exists
    `;

    return rows[0]?.exists ?? false;
  }

  async getPlatformSettings(tx: TransactionClient = this.prisma) {
    const settings = await tx.platform_settings.findUnique({
      where: { id: 1 },
    });

    if (!settings) {
      throw new BadRequestError('Platform settings are not configured');
    }

    return settings;
  }

  async createBookingHold(
    options: {
      customerId: string;
      roomId: string;
      propertyId: string;
      checkIn: Date;
      checkOut: Date;
      guestCount: number;
      customerNotes: string | null;
      bookingMode: string;
      price: BookingPrice;
      nights: number;
      monthlyPriceKrw: number;
    },
    tx: TransactionClient = this.prisma,
  ): Promise<bookings> {
    const settings = await this.getPlatformSettings(tx);
    const now = new Date();

    let status: booking_status;
    let bookingTypeValue: booking_type;
    let holdExpiresAt: Date | null;

    if (options.bookingMode === booking_mode.instant) {
      status = booking_status.pending_payment;
      bookingTypeValue = booking_type.instant;
      holdExpiresAt = new Date(
        now.getTime() + settings.hold_ttl_minutes * 60_000,
      );
    } else {
      status = booking_status.requested;
      bookingTypeValue = booking_type.request;
      holdExpiresAt = null;
    }

    try {
      const booking = await tx.bookings.create({
        data: {
          customer_id: options.customerId,
          room_id: options.roomId,
          property_id: options.propertyId,
          check_in: options.checkIn,
          check_out: options.checkOut,
          guest_count: options.guestCount,
          status,
          booking_type: bookingTypeValue,
          hold_expires_at: holdExpiresAt,
          customer_notes: options.customerNotes,
        },
      });

      await tx.booking_price_snapshots.create({
        data: {
          booking_id: booking.id,
          rent_krw: options.price.rentKrw,
          service_fee_krw: options.price.serviceFeeKrw,
          utilities_krw: 0,
          total_krw: options.price.totalKrw,
          pricing_version: options.price.pricingVersion,
          nightly_breakdown: [
            {
              nights: options.nights,
              monthly_price_krw: options.monthlyPriceKrw,
              rent_krw: options.price.rentKrw,
              service_fee_krw: options.price.serviceFeeKrw,
            },
          ],
        },
      });

      return booking;
    } catch (error) {
      rethrowBookingOverlap(error);
    }
  }

  async listCustomerBookings(
    customerId: string,
    tx: TransactionClient = this.prisma,
  ): Promise<BookingListRow[]> {
    const rows = await tx.$queryRaw<BookingListRowRecord[]>`
      ${Prisma.raw(BOOKING_LIST_SQL)}
      where b.customer_id = ${customerId}::uuid
      order by b.created_at desc
    `;

    return rows.map(mapBookingListRow);
  }

  async getBooking(
    bookingId: string,
    tx: TransactionClient = this.prisma,
  ): Promise<bookings | null> {
    return tx.bookings.findUnique({
      where: { id: bookingId },
    });
  }

  async getBookingDetailRow(
    bookingId: string,
    tx: TransactionClient = this.prisma,
  ): Promise<BookingListRow | null> {
    const rows = await tx.$queryRaw<BookingListRowRecord[]>`
      ${Prisma.raw(BOOKING_LIST_SQL)}
      where b.id = ${bookingId}::uuid
    `;

    const row = rows[0];
    return row ? mapBookingListRow(row) : null;
  }

  async cancelBooking(
    bookingId: string,
    customerId: string,
    tx: TransactionClient = this.prisma,
  ): Promise<bookings | null> {
    const booking = await tx.bookings.findFirst({
      where: {
        id: bookingId,
        customer_id: customerId,
        status: {
          in: [booking_status.requested, booking_status.pending_payment],
        },
      },
    });

    if (!booking) {
      return null;
    }

    return tx.bookings.update({
      where: { id: bookingId },
      data: {
        status: booking_status.cancelled,
        cancelled_at: new Date(),
      },
    });
  }

  async approveBooking(
    bookingId: string,
    approvedBy: string,
    tx: TransactionClient = this.prisma,
  ): Promise<bookings | null> {
    const booking = await this.getBooking(bookingId, tx);

    if (!booking || booking.status !== booking_status.requested) {
      return null;
    }

    if (
      await this.roomHasBookingConflict(
        booking.room_id,
        booking.check_in,
        booking.check_out,
        tx,
      )
    ) {
      throw new ConflictError(
        'Selected dates conflict with an existing booking hold',
      );
    }

    const settings = await this.getPlatformSettings(tx);
    const now = new Date();

    try {
      return await tx.bookings.update({
        where: { id: bookingId },
        data: {
          status: booking_status.pending_payment,
          hold_expires_at: new Date(
            now.getTime() + settings.hold_ttl_minutes * 60_000,
          ),
          approved_at: now,
          approved_by: approvedBy,
        },
      });
    } catch (error) {
      rethrowBookingOverlap(error);
    }
  }

  async rejectBooking(
    bookingId: string,
    tx: TransactionClient = this.prisma,
  ): Promise<bookings | null> {
    const booking = await tx.bookings.findFirst({
      where: {
        id: bookingId,
        status: booking_status.requested,
      },
    });

    if (!booking) {
      return null;
    }

    return tx.bookings.update({
      where: { id: bookingId },
      data: {
        status: booking_status.rejected,
      },
    });
  }
}

function mapBookingListRow(row: BookingListRowRecord): BookingListRow {
  return {
    id: row.id,
    customerId: row.customer_id,
    status: row.status,
    bookingType: row.booking_type,
    checkIn: row.check_in,
    checkOut: row.check_out,
    guestCount: row.guest_count,
    holdExpiresAt: row.hold_expires_at,
    createdAt: row.created_at,
    customerNotes: row.customer_notes,
    propertyId: row.property_id,
    roomId: row.room_id,
    propertyTitle: row.property_title,
    district: row.district,
    roomName: row.room_name,
    rentKrw: row.rent_krw,
    serviceFeeKrw: row.service_fee_krw,
    totalKrw: row.total_krw,
    pricingVersion: row.pricing_version,
  };
}

function daysBetween(checkIn: Date, checkOut: Date): number {
  return Math.round((checkOut.getTime() - checkIn.getTime()) / 86_400_000);
}

File: apps/backend-java/src/main/java/com/housingplatform/bookings/BookingRepository.java
package com.housingplatform.bookings;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.housingplatform.auth.error.BadRequestException;
import com.housingplatform.auth.error.ConflictException;
import com.housingplatform.auth.error.NotFoundException;
import com.housingplatform.persistence.entity.Booking;
import com.housingplatform.persistence.enums.BookingStatus;
import com.housingplatform.persistence.enums.BookingType;
import jakarta.persistence.EntityManager;
import jakarta.persistence.NoResultException;
import jakarta.persistence.PersistenceContext;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import org.hibernate.exception.ConstraintViolationException;
import org.springframework.stereotype.Repository;

@Repository
public class BookingRepository {

  @PersistenceContext private EntityManager entityManager;

  private final ObjectMapper objectMapper;

  public BookingRepository(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  public record BookingInputs(
      UUID roomId,
      UUID propertyId,
      String bookingMode,
      int minStayNights,
      int monthlyPriceKrw,
      int maxOccupancy,
      int nights) {}

  public record BookingPrice(int rentKrw, int serviceFeeKrw, int totalKrw, String pricingVersion) {}

  public record BookingListRow(
      UUID id,
      UUID customerId,
      BookingStatus status,
      BookingType bookingType,
      LocalDate checkIn,
      LocalDate checkOut,
      int guestCount,
      OffsetDateTime holdExpiresAt,
      OffsetDateTime createdAt,
      String customerNotes,
      UUID propertyId,
      UUID roomId,
      String propertyTitle,
      String district,
      String roomName,
      int rentKrw,
      int serviceFeeKrw,
      int totalKrw,
      String pricingVersion) {}

  public BookingInputs validateBookingInputs(
      UUID roomId, LocalDate checkIn, LocalDate checkOut, int guestCount) {
    if (!checkOut.isAfter(checkIn)) {
      throw new BadRequestException("checkOut must be after checkIn");
    }
    if (guestCount <= 0) {
      throw new BadRequestException("guestCount must be positive");
    }

    Object[] roomRow;
    try {
      roomRow =
          (Object[])
              entityManager
                  .createNativeQuery(
                      """
                      select r.id, r.property_id, r.max_occupancy, r.monthly_price_krw, r.available_from
                      from public.rooms r
                      where r.id = :roomId
                        and r.deleted_at is null
                        and r.status = 'available'
                      """)
                  .setParameter("roomId", roomId)
                  .getSingleResult();
    } catch (NoResultException exception) {
      throw new NotFoundException("Room is not available");
    }

    UUID resolvedRoomId = (UUID) roomRow[0];
    UUID propertyId = (UUID) roomRow[1];
    int maxOccupancy = ((Number) roomRow[2]).intValue();
    int monthlyPriceKrw = ((Number) roomRow[3]).intValue();
    LocalDate availableFrom = toLocalDate(roomRow[4]);

    Object[] propertyRow;
    try {
      propertyRow =
          (Object[])
              entityManager
                  .createNativeQuery(
                      """
                      select p.id, p.booking_mode::text, p.min_stay_nights
                      from public.properties p
                      where p.id = :propertyId
                        and p.deleted_at is null
                        and p.status = 'published'
                      """)
                  .setParameter("propertyId", propertyId)
                  .getSingleResult();
    } catch (NoResultException exception) {
      throw new NotFoundException("Property is not published");
    }

    String bookingMode = (String) propertyRow[1];
    int minStayNights = ((Number) propertyRow[2]).intValue();
    int nights = (int) java.time.temporal.ChronoUnit.DAYS.between(checkIn, checkOut);

    if (nights < minStayNights) {
      throw new BadRequestException("Stay must be at least " + minStayNights + " nights");
    }
    if (guestCount > maxOccupancy) {
      throw new BadRequestException("Room supports up to " + maxOccupancy + " guests");
    }
    if (availableFrom != null && availableFrom.isAfter(checkIn)) {
      throw new BadRequestException("Room is not available from the selected check-in date");
    }

    return new BookingInputs(
        resolvedRoomId, propertyId, bookingMode, minStayNights, monthlyPriceKrw, maxOccupancy, nights);
  }

  public BookingPrice calculateBookingPrice(int monthlyPriceKrw, int nights) {
    try {
      Object[] row =
          (Object[])
              entityManager
                  .createNativeQuery(
                      """
                      select rent_krw, service_fee_krw, total_krw, pricing_version
                      from public.calculate_booking_price(:monthlyPriceKrw, :nights)
                      """)
                  .setParameter("monthlyPriceKrw", monthlyPriceKrw)
                  .setParameter("nights", nights)
                  .getSingleResult();
      return new BookingPrice(
          ((Number) row[0]).intValue(),
          ((Number) row[1]).intValue(),
          ((Number) row[2]).intValue(),
          (String) row[3]);
    } catch (NoResultException exception) {
      throw new BadRequestException("Unable to calculate booking price");
    }
  }

  public boolean roomHasBookingConflict(UUID roomId, LocalDate checkIn, LocalDate checkOut) {
    Object result =
        entityManager
            .createNativeQuery(
                """
                select public.room_has_booking_conflict(:roomId, :checkIn, :checkOut)
                """)
            .setParameter("roomId", roomId)
            .setParameter("checkIn", checkIn)
            .setParameter("checkOut", checkOut)
            .getSingleResult();
    return Boolean.TRUE.equals(result);
  }

  public int getHoldTtlMinutes() {
    try {
      Number value =
          (Number)
              entityManager
                  .createNativeQuery(
                      """
                      select hold_ttl_minutes from public.platform_settings where id = 1
                      """)
                  .getSingleResult();
      return value.intValue();
    } catch (NoResultException exception) {
      throw new BadRequestException("Platform settings are not configured");
    }
  }

  public UUID createBookingHold(
      UUID customerId,
      BookingInputs inputs,
      LocalDate checkIn,
      LocalDate checkOut,
      int guestCount,
      String customerNotes,
      BookingPrice price) {
    int holdTtlMinutes = getHoldTtlMinutes();
    OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);

    String status;
    String bookingType;
    OffsetDateTime holdExpiresAt;
    if ("instant".equals(inputs.bookingMode())) {
      status = "pending_payment";
      bookingType = "instant";
      holdExpiresAt = now.plusMinutes(holdTtlMinutes);
    } else {
      status = "requested";
      bookingType = "request";
      holdExpiresAt = null;
    }

    UUID bookingId;
    try {
      bookingId =
          (UUID)
              entityManager
                  .createNativeQuery(
                      """
                      insert into public.bookings (
                        customer_id, room_id, property_id, check_in, check_out, guest_count,
                        status, booking_type, hold_expires_at, customer_notes
                      ) values (
                        :customerId, :roomId, :propertyId, :checkIn, :checkOut, :guestCount,
                        cast(:status as booking_status), cast(:bookingType as booking_type),
                        :holdExpiresAt, :customerNotes
                      )
                      returning id
                      """)
                  .setParameter("customerId", customerId)
                  .setParameter("roomId", inputs.roomId())
                  .setParameter("propertyId", inputs.propertyId())
                  .setParameter("checkIn", checkIn)
                  .setParameter("checkOut", checkOut)
                  .setParameter("guestCount", guestCount)
                  .setParameter("status", status)
                  .setParameter("bookingType", bookingType)
                  .setParameter("holdExpiresAt", holdExpiresAt)
                  .setParameter("customerNotes", customerNotes)
                  .getSingleResult();
    } catch (RuntimeException exception) {
      if (isOverlapConflict(exception)) {
        throw new ConflictException("Selected dates conflict with an existing booking hold");
      }
      throw exception;
    }

    try {
      String nightlyBreakdown = buildNightlyBreakdown(inputs, price);
      entityManager
          .createNativeQuery(
              """
              insert into public.booking_price_snapshots (
                booking_id, rent_krw, service_fee_krw, utilities_krw, total_krw,
                pricing_version, nightly_breakdown
              ) values (
                :bookingId, :rentKrw, :serviceFeeKrw, 0, :totalKrw,
                :pricingVersion, cast(:nightlyBreakdown as jsonb)
              )
              """)
          .setParameter("bookingId", bookingId)
          .setParameter("rentKrw", price.rentKrw())
          .setParameter("serviceFeeKrw", price.serviceFeeKrw())
          .setParameter("totalKrw", price.totalKrw())
          .setParameter("pricingVersion", price.pricingVersion())
          .setParameter("nightlyBreakdown", nightlyBreakdown)
          .executeUpdate();
      entityManager.flush();
    } catch (RuntimeException exception) {
      if (isOverlapConflict(exception)) {
        throw new ConflictException("Selected dates conflict with an existing booking hold");
      }
      throw exception;
    }

    return bookingId;
  }

  public List<BookingListRow> listCustomerBookings(UUID customerId) {
    return queryBookingListRows(
        """
        where b.customer_id = :customerId
        order by b.created_at desc
        """,
        Map.of("customerId", customerId));
  }

  public Optional<BookingListRow> findBookingDetailRow(UUID bookingId) {
    List<BookingListRow> rows =
        queryBookingListRows(
            """
            where b.id = :bookingId
            """,
            Map.of("bookingId", bookingId));
    return rows.stream().findFirst();
  }

  public Optional<Booking> findBooking(UUID bookingId) {
    return entityManager
        .createQuery(
            """
            select b from Booking b where b.id = :bookingId
            """,
            Booking.class)
        .setParameter("bookingId", bookingId)
        .getResultStream()
        .findFirst();
  }

  public Booking cancelBooking(UUID bookingId, UUID customerId) {
    int updated =
        entityManager
            .createNativeQuery(
                """
                update public.bookings
                set status = 'cancelled',
                    cancelled_at = timezone('utc', now()),
                    updated_at = timezone('utc', now())
                where id = :bookingId
                  and customer_id = :customerId
                  and status in ('requested', 'pending_payment')
                """)
            .setParameter("bookingId", bookingId)
            .setParameter("customerId", customerId)
            .executeUpdate();
    if (updated == 0) {
      return null;
    }
    entityManager.flush();
    return findBooking(bookingId).orElse(null);
  }

  public Booking approveBooking(UUID bookingId, UUID approvedBy) {
    Booking booking =
        findBooking(bookingId).orElse(null);
    if (booking == null || booking.getStatus() != BookingStatus.requested) {
      return null;
    }

    if (roomHasBookingConflict(booking.getRoomId(), booking.getCheckIn(), booking.getCheckOut())) {
      throw new ConflictException("Selected dates conflict with an existing booking hold");
    }

    int holdTtlMinutes = getHoldTtlMinutes();
    OffsetDateTime holdExpiresAt = OffsetDateTime.now(ZoneOffset.UTC).plusMinutes(holdTtlMinutes);

    try {
      int updated =
          entityManager
              .createNativeQuery(
                  """
                  update public.bookings
                  set status = 'pending_payment',
                      hold_expires_at = :holdExpiresAt,
                      approved_at = timezone('utc', now()),
                      approved_by = :approvedBy,
                      updated_at = timezone('utc', now())
                  where id = :bookingId and status = 'requested'
                  """)
              .setParameter("bookingId", bookingId)
              .setParameter("holdExpiresAt", holdExpiresAt)
              .setParameter("approvedBy", approvedBy)
              .executeUpdate();
      if (updated == 0) {
        return null;
      }
      entityManager.flush();
    } catch (RuntimeException exception) {
      if (isOverlapConflict(exception)) {
        throw new ConflictException("Selected dates conflict with an existing booking hold");
      }
      throw exception;
    }

    return findBooking(bookingId).orElse(null);
  }

  public Booking rejectBooking(UUID bookingId) {
    int updated =
        entityManager
            .createNativeQuery(
                """
                update public.bookings
                set status = 'rejected',
                    updated_at = timezone('utc', now())
                where id = :bookingId and status = 'requested'
                """)
            .setParameter("bookingId", bookingId)
            .executeUpdate();
    if (updated == 0) {
      return null;
    }
    entityManager.flush();
    return findBooking(bookingId).orElse(null);
  }

  public void notifyBookingRequest(UUID bookingId) {
    entityManager
        .createNativeQuery(
            """
            select public.notify_booking_request(b)
            from public.bookings b
            where b.id = :bookingId
            """)
        .setParameter("bookingId", bookingId)
        .getSingleResult();
  }

  public void notifyBookingConfirmed(UUID bookingId) {
    entityManager
        .createNativeQuery(
            """
            select public.notify_booking_confirmed(b)
            from public.bookings b
            where b.id = :bookingId
            """)
        .setParameter("bookingId", bookingId)
        .getSingleResult();
  }

  public void notifyBookingRejected(UUID bookingId) {
    entityManager
        .createNativeQuery(
            """
            select public.notify_booking_rejected(b)
            from public.bookings b
            where b.id = :bookingId
            """)
        .setParameter("bookingId", bookingId)
        .getSingleResult();
  }

  @SuppressWarnings("unchecked")
  private List<BookingListRow> queryBookingListRows(String whereClause, Map<String, Object> params) {
    var query =
        entityManager.createNativeQuery(
            """
            select
              b.id,
              b.customer_id,
              b.status::text,
              b.booking_type::text,
              b.check_in,
              b.check_out,
              b.guest_count,
              b.hold_expires_at,
              b.created_at,
              b.customer_notes,
              b.property_id,
              b.room_id,
              p.title,
              p.district,
              r.name,
              s.rent_krw,
              s.service_fee_krw,
              s.total_krw,
              s.pricing_version
            from public.bookings b
            join public.properties p on p.id = b.property_id
            join public.rooms r on r.id = b.room_id
            join public.booking_price_snapshots s on s.booking_id = b.id
            """
                + whereClause);

    params.forEach(query::setParameter);
    List<Object[]> rows = query.getResultList();
    List<BookingListRow> results = new ArrayList<>();
    for (Object[] row : rows) {
      results.add(
          new BookingListRow(
              (UUID) row[0],
              (UUID) row[1],
              BookingStatus.valueOf((String) row[2]),
              BookingType.valueOf((String) row[3]),
              toLocalDate(row[4]),
              toLocalDate(row[5]),
              ((Number) row[6]).intValue(),
              (OffsetDateTime) row[7],
              (OffsetDateTime) row[8],
              (String) row[9],
              (UUID) row[10],
              (UUID) row[11],
              (String) row[12],
              (String) row[13],
              (String) row[14],
              ((Number) row[15]).intValue(),
              ((Number) row[16]).intValue(),
              ((Number) row[17]).intValue(),
              (String) row[18]));
    }
    return results;
  }

  private String buildNightlyBreakdown(BookingInputs inputs, BookingPrice price) {
    Map<String, Object> entry = new LinkedHashMap<>();
    entry.put("nights", inputs.nights());
    entry.put("monthly_price_krw", inputs.monthlyPriceKrw());
    entry.put("rent_krw", price.rentKrw());
    entry.put("service_fee_krw", price.serviceFeeKrw());
    try {
      return objectMapper.writeValueAsString(List.of(entry));
    } catch (JsonProcessingException exception) {
      throw new IllegalStateException("Unable to serialize nightly breakdown", exception);
    }
  }

  private static LocalDate toLocalDate(Object value) {
    if (value == null) {
      return null;
    }
    if (value instanceof LocalDate localDate) {
      return localDate;
    }
    if (value instanceof java.sql.Date sqlDate) {
      return sqlDate.toLocalDate();
    }
    return LocalDate.parse(value.toString());
  }

  private static boolean isOverlapConflict(Throwable exception) {
    Throwable current = exception;
    while (current != null) {
      if (current instanceof ConstraintViolationException constraint) {
        String message = constraint.getSQLException().getMessage();
        if (message != null) {
          String lowered = message.toLowerCase();
          if (lowered.contains("bookings_no_overlap") || lowered.contains("exclusion")) {
            return true;
          }
        }
      }
      String message = current.getMessage();
      if (message != null) {
        String lowered = message.toLowerCase();
        if (lowered.contains("bookings_no_overlap") || lowered.contains("exclusion")) {
          return true;
        }
      }
      current = current.getCause();
    }
    return false;
  }
}
